### DISCUSSION TITLE
How and where are Bookmarks stored

### DISCUSSION THREAD

**Discussion by bernd-wechner** [smplayer-dev/smplayer#147] [Q&A]
I'm using Portable SMplayer 21.1.0 on MS Windows 10.  I make bookmarks in a video and I can use them. That rocks. SMplayer remembers these bookmarks between runs. So it must be remembering them somehow.

How? And where?

The only clue I have found is Ubuntu specific:

https://askubuntu.com/questions/817780/smplayer-bookmarks-backup

but mentions `favorites.m3u8`. I see such a file in my SMplayer folder but it is empty. Well contains:

```
#EXTM3U

```

Yet, SMplayer remembers the bookmarks I have set. I search my entire disk for other copies of `favorites.m3u8`, there are none. I search the registry for all smplayer references, and find no clues. 

I would like to back these up, I'd also like to be able to read them. How and where does SMplayer remember these bookmarks on MS Windows 10?

**Comment by smplayer-dev** [comment]
They are in ini files inside the `file_settings` directory.

**Reply by bernd-wechner** [reply]
Thanks kindly. I do find some clues in that folder. A single ini file.

`2a53cb67436f4e21.ini`

and it does contain bookmarks in the form:
```
bookmarks\1\time=0
bookmarks\1\name=
bookmarks\2\time=9
bookmarks\2\name=Start 1
bookmarks\3\time=48
bookmarks\size=3
bookmarks\3\name=End 1
```

The UI clearly ties these marks to a specific video file. I can't see how that happens. See no candidates in the ini file. And wonder if it's name perhaps is a hash of the full path of the video file name? And if so, by what method? Or if the binding happens some other way?

**Comment by smplayer-dev** [comment]
Those setting files are created by [FileSettingsHash](https://github.com/smplayer-dev/smplayer/blob/master/src/filesettingshash.h).
There's the possibility to use only one ini file for all files (preferences -> general -> store settings in ->  one ini file), but that file can grow too much and so it can increase the time to load/save it.
